### ISSUE TITLE
Inconsistent defaults for class libraries vs. plugin scenario

### ISSUE THREAD

**Issue by nguerrera** [dotnet/sdk#2662]
Runtime is adding a feature to load components with deps.json.

However, IIRC, libraries will not copy their nuget refs by default in the build and publish unification thus far. So folks might expect they can just deploy their plugin build output, but it will be missing dependencies.

We need to think about this. Opting in for libs makes sense to me on the one hand because it is common to have a large sln full of libs that will never deploy as their own thing and it could lead to a lot of extra copying. On the other hand, we generate deps.json for libs by default yet it isn't needed in this common case I'm describing and is far from free to generate. I've wanted to change that. Maybe deps.json should also be behind HasRuntimeOutput being opted in, similar in spirit to binding redirects. The guidance would then be to set it in plugin project.

The other alternative is to default both copy local of nuget dependencies and deps.json to on for libraries  Bottom line is that these two should go hand in hand.

**Comment by nguerrera** [comment]
@peterhuene @vitek-karas

**Comment by eerhardt** [comment]
The `.deps.json` file generated by a library is _very_ incorrect w.r.t. what gets used at runtime. And it shouldn't be used at runtime. I believe it was a mistake to create it in the first place.

The reason it is wrong is that building a library doesn't know which actual referenced assembly is going to be used at runtime.

For example, say I have a `netstandard2.0` library and it depends on a NuGet package with the following structure:

```
- ref
  - netstandard.2.0
    - foo.dll
- lib
  - netcoreapp2.0
    - foo.dll
  - netfx45
    - foo.dll
```

When the application is running on the desktop .NET Framework, `lib\netfx45\foo.dll` will be used.  When the application is running on .NET Core 2.0+, `lib\netcoreapp2.0\foo.dll` will be used.

There is no way of knowing this ahead of time in the `.deps.json` file when building the library. Only once an executable TFM is picked, will this information be available.

/cc @ericstj

**Comment by eerhardt** [comment]
One "idea" to get this scenario to work is require the developer pick the executable TFM when building/publishing the above library. If they want that library to run on `netfx45`, they need to build/publish the library for `netfx45`. Similar for `netcoreappX.Y`.

**Comment by vitek-karas** [comment]
We're kind of relying on the `.deps.json` to enable plugin like scenarios to work (see the proposed API here: dotnet/corefx#33165). That said I'm still trying to figure out what are the recommendations for plugin authors to follow. One such recommendation could be that they have to build it as a `netcoreapp2.0` classlib (not `netstandard`).

**Comment by rseanhall** [comment]
Isn't this issue resolved with the `EnableDynamicLoading` property from https://github.com/dotnet/sdk/pull/3305?

**Comment by vitek-karas** [comment]
@rseanhall is right, this is basically solved by #3305.
